src/gui/widgets/tab_container.hpp

```diff
@@ -0,0 +1,147 @@
+/*
+	Copyright (C) 2024
+	by Subhraman Sarkar (babaissarkar) <suvrax@gmail.com>
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#pragma once
+
+#include "gui/widgets/container_base.hpp"
+#include "gui/widgets/listbox.hpp"
+#include "gui/core/window_builder.hpp"
+
+namespace gui2
+{
+
+namespace implementation
+{
+struct builder_tab_container;
+}
+
+/**
+ * A container widget that shows one of its pages of widgets
+ * depending on which tab the user clicked.
+ */
+class tab_container : public container_base
+{
+	friend struct implementation::builder_tab_container;
+
+public:
+	explicit tab_container(const implementation::builder_tab_container& builder);
+
+	/** See @ref container_base::set_self_active. */
+	virtual void set_self_active(const bool active) override;
+
+	/***** ***** ***** setters / getters for members ***** ****** *****/
+
+	/** See @ref styled_widget::get_active. */
+	virtual bool get_active() const override;
+
+	/** See @ref styled_widget::get_state. */
+	virtual unsigned get_state() const override;
+
+	bool can_wrap() const override;
+
+	void add_tab_entry(const widget_data row);
```

Comment:
Judging by the implementation, this should be private, as calling it directly puts the UI widgets out of sync with `list_items_`.

Reply:
made private.